pkg/apb/executor.go

```diff
@@ -32,11 +32,74 @@ import (
 	"k8s.io/api/core/v1"
 )
 
-// ExecuteApb - Runs an APB Action with a provided set of inputs
-func ExecuteApb(action string,
-	spec *Spec,
-	context *Context,
-	p *Parameters) (ExecutionContext, error) {
+// Executor - Main struct used for running APBs. Manages their lifecycle.
+type Executor struct {
+	extractedCredentials *ExtractedCredentials
+	podName              string
+	lastStatus           StatusMessage
+	statusChan           chan StatusMessage
+}
+
+// NewExecutor - Creates a new Executor for running an APB.
+func NewExecutor() *Executor {
+	exec := &Executor{
+		statusChan: make(chan StatusMessage),
+		lastStatus: StatusMessage{State: StateNotYetStarted},
+	}
+	return exec
+}
+
+// GetPodName- Returns the name of the pod running the APB
+func (e *Executor) GetPodName() string {
+	return e.podName
+}
+
+// GetLastStatus - Returns the last known status of the APB
+func (e *Executor) GetLastStatus() StatusMessage {
+	return e.lastStatus
+}
+
+// ExtractedCredentials - Credentials extracted from the APB while running,
+// if they were discovered.
+func (e *Executor) GetExtractedCredentials() *ExtractedCredentials {
+	return e.extractedCredentials
+}
+
+func (e *Executor) start() {
```

Comment:
They're really lifecycle hooks that need to be called from the action methods to alert the `Executor` to the current state of what's going on (and it abstracts the fact that messages are being sent). I originally named `updateDescriptor` to something like `sendDescriptorMessage`, but I felt like the latter was a less abstract name. I want to call high level hooks to alert the executor to triggers in the cycle without saying, "hey send this message", because it's better if the actions don't even care about that. They just say "hey, this thing happened, do what you would like with this info".

Reply:
I'll also say that I tried to avoid imposing additional requirements on callers since I expect these to turn into the primary methods that people will use with `libapb`. The worst of it is to ask folks to instantiate an `Executor` with `NewExecutor`; I've personally seen us on this path for a while, so I'm happy to finally get this down on paper, so to speak.

Reply:
Do we think there is a better name for `Executor`? I'm still hesitant to name these something like "sendX" or other related messaging concept. They're internal lifecycle hooks intended to be called by the internal APB code, and do more than send a message over the buffer (not much today, but tomorrow they might).

> also would give an opportunity for this to be an interface that could be overridden for tests or other uses if we exposed it via a constructor.

Definitely interested in this since decent test coverage is the biggest thing missing right now. I don't think callers should have to worry about injecting that via ctor though. A sensible default that can get overriden by test code I think makes sense, given this is an internal detail.

Reply:
> Definitely interested in this since decent test coverage is the biggest thing missing right now. I don't think callers should have to worry about injecting that via ctor though. A sensible default that can get overriden by test code I think makes sense, given this is an internal detail.

Yep agree no need to expose as long as it can be overridden in test code